### PR TITLE
Add tooling to make customization of DB Connection possible

### PR DESCRIPTION
**Pull request by alukach**
**Description:**

This PR updates the way in which the pgSTAC extension connects with its underlying database.  This was done to support the possibility of users running custom connection setup/teardown steps for each request.  This allows us to do something like the following when setting up a pgSTAC-based STAC API:

```diff
diff --git a/stac_fastapi/pgstac/stac_fastapi/pgstac/app.py b/stac_fastapi/pgstac/stac_fastapi/pgstac/app.py
index 8a33424..1033f11 100644
--- a/stac_fastapi/pgstac/stac_fastapi/pgstac/app.py
+++ b/stac_fastapi/pgstac/stac_fastapi/pgstac/app.py
@@ -6,8 +6,14 @@ If the variable is not set, enables all extensions.
 """
 
 import os
+from contextlib import asynccontextmanager
+from typing import Literal
 
+from asyncpg import exceptions, Connection
+from buildpg import render
+from fastapi import HTTPException, Request
 from fastapi.responses import ORJSONResponse
+from fastapi.security import HTTPBasic
 
 from stac_fastapi.api.app import StacApi
 from stac_fastapi.api.models import create_get_request_model, create_post_request_model
@@ -22,7 +28,7 @@ from stac_fastapi.extensions.core import (
 from stac_fastapi.extensions.third_party import BulkTransactionExtension
 from stac_fastapi.pgstac.config import Settings
 from stac_fastapi.pgstac.core import CoreCrudClient
-from stac_fastapi.pgstac.db import close_db_connection, connect_to_db
+from stac_fastapi.pgstac.db import close_db_connection, connect_to_db, get_connection
 from stac_fastapi.pgstac.extensions import QueryExtension
 from stac_fastapi.pgstac.extensions.filter import FiltersClient
 from stac_fastapi.pgstac.transactions import BulkTransactionsClient, TransactionsClient
@@ -65,10 +71,57 @@ api = StacApi(
 app = api.app
 
 
+auth = HTTPBasic()
+
+
+async def set_db_user(conn: Connection, request: Request) -> None:
+    # Require & parse basic auth
+    credentials = await auth(request)
+
+    # Validate credentials...
+    q, p = render(
+        f"""
+        SELECT authenticate_user(:username::text, :password::text)
+        """,
+        username=credentials.username,
+        password=credentials.password,
+    )
+    try:
+        await conn.execute(q, *p)
+    except exceptions.RaiseError as e:
+        raise HTTPException(status_code=401, detail="Bad credentials") from e
+
+    # Set user in DB...
+    q, p = render(
+        f"""
+        SELECT set_config('api.username', :username::text, false)
+        """,
+        username=credentials.username,
+    )
+    await conn.execute(q, *p)
+
+
+@asynccontextmanager
+async def custom_get_connection(
+    request: Request,
+    readwrite: Literal["r", "w"],
+):
+    """An example of customizing the connection getter"""
+    async with get_connection(request, readwrite) as conn:
+        if readwrite == "w":
+            await set_db_user(conn, request)
+
+        # Simple demo to validate we now know who is authenticated
+        current_user = await conn.fetchval("SELECT current_setting('api.username', true)")
+        print(f"Running DB command as {current_user!r}")
+        
+        yield conn
+
+
 @app.on_event("startup")
 async def startup_event():
     """Connect to database on startup."""
-    await connect_to_db(app)
+    await connect_to_db(app, custom_get_connection)
 
 
 @app.on_event("shutdown")
```

**PR Checklist:**

- [ ] Code is formatted and linted (run `pre-commit run --all-files`)
- [ ] Tests pass (run `make test`)
- [ ] Documentation has been updated to reflect changes, if applicable, and docs build successfully (run `make docs`)
- [ ] Changes are added to the [CHANGELOG](https://github.com/stac-utils/stac-fastapi/blob/master/CHANGES.md).

### CODE REVIEW

**Review by vincentsarago — APPROVED** [review]

File: stac_fastapi/pgstac/stac_fastapi/pgstac/core.py [hunk 1]
```diff
@@ -38,13 +37,11 @@
...
 
-    async def all_collections(self, **kwargs) -> Collections:
+    async def all_collections(self, request: Request, **kwargs) -> Collections:
         """Read all collections from the database."""
-        request: Request = kwargs["request"]
         base_url = get_base_url(request)
-        pool = request.app.state.readpool
 
-        async with pool.acquire() as conn:
+        async with request.app.state.get_connection(request, "r") as conn:
```
Comment (vincentsarago) [inline stac_fastapi/pgstac/stac_fastapi/pgstac/core.py:44]: To be honest it feels kinda weird to have to pass `request` to an function we get from the request object but sadly I don't think we have any choice 😭

**Comment by gadomski** [comment]
Just a note that the CI failures are due to non-smart configuration of the GHCR updates (by me), and should be fixed by https://github.com/stac-utils/stac-fastapi/pull/535 (which will close https://github.com/stac-utils/stac-fastapi/issues/534).

**Comment by alukach** [comment]
Alright @bitner & others, I believe this is now good to go.  Testing is failing due to the issue described in https://github.com/stac-utils/pgstac/issues/177, but that should be fixed when we upgrade to pgSTAC 0.7.5 (via fix from https://github.com/stac-utils/pgstac/pull/178).

I added one simple test to validate that the customized DB connection can be used to run logic with API requests: https://github.com/stac-utils/stac-fastapi/blob/befe6ea676ffa4fcab583a57313f1efb5768c476/stac_fastapi/pgstac/tests/clients/test_postgres.py#L182-L212

**Review by bitner — APPROVED** [review]
Looks good! Thank you @alukach. @gadomski I just want to make sure that you see this too as far as it relates to the backend separation PR.

**Comment by gadomski** [comment]
Sounds good, thanks for the heads up. I'll merge this as is, and then open an issue to backport it to https://github.com/stac-utils/stac-fastapi-pgstac/.

**Comment by gadomski** [comment]
https://github.com/stac-utils/stac-fastapi-pgstac/issues/19

**Pull request merged**